crates/sui-core/src/checkpoints/mod.rs

```diff
@@ -378,23 +388,44 @@ impl CheckpointStore {
         epoch: EpochId,
         sequence_number: CheckpointSequenceNumber,
         transactions: impl Iterator<Item = &'a ExecutionDigests> + Clone,
-        effects_store: impl CausalOrder,
+        effects_store: impl CausalOrder + EffectsStore,
         next_epoch_committee: Option<Committee>,
     ) -> SuiResult {
         let previous_digest = self.get_prev_checkpoint_digest(sequence_number)?;
 
         // Create a causal order of all transactions in the checkpoint.
         let ordered_contents = CheckpointContents::new_with_causally_ordered_transactions(
             effects_store
-                .get_complete_causal_order(transactions, self)?
+                .get_complete_causal_order(transactions.clone(), self)?
                 .into_iter(),
         );
 
+        let (storage_charges, computation_charges, storage_rebates): (
+            Vec<u64>,
+            Vec<u64>,
+            Vec<u64>,
+        ) = effects_store
+            .get_effects(transactions)?
```

Comment:
This will lead to another read of all effects, which is not very efficient. Ideally we should reuse the effects read inside get_complete_causal_order

Reply:
I understand your concern. I didn't do this inside `get_complete_causal_order` because `get_complete_causal_order` is part of the `CausalOrder` interface, which doesn't have the functionality to get transaction effects. One thing I think we can do here is to combine `CausalOrder` and `EffectsStore` traits. But from the comment it seems that we want to keep the interfaces separate? https://github.com/MystenLabs/sui/blob/main/crates/sui-core/src/checkpoints/causal_order_effects.rs#L20-L22
Actually I'm not sure why `CausalOrder` exists when we already have `EffectsStore::causal_order_from_effects`. What do you think? @lxfind

Reply:
You can put it inside `causal_order_from_effects`.
I think this was done to make separated testing easier. But yeah I could see that maybe they could just be merged given that we always implement both anyway. cc @gdanezis who authored this.

Reply:
> You can put it inside causal_order_from_effects.

Yes and that would be essentially merging the two traits and rendering `CausalOrder` useless, at least in our current code. I'll go ahead and do that without removing the CausalOrder trait from the code, until George gets back.